text/0055-nested-multi-sigs.md

```diff
@@ -0,0 +1,101 @@
+
+We propose to add a third field to the struct, `multi_signers`, that allows including other `SignerSet`s. Signature validation code will be changed to look for matches over both `individual_signers` and `multi_signers`, where a match over a multi-signer is considered as a single signature when counting against the threshold.
+
+We will make minor changes to the `tokens.json` file format and the `mc-consensus-mint-client` tool to allow specifying nested signer sets.
+
+Since we are only adding a field, this change is backwards compatible when it comes to the protobuf representation of `SignerSet`s. . However, `SignerSet`s make their way into blocks (they are included in `MintConfig`s and `ValidatedMintConfigTx`s and those go into `BlockContents`. This means we will need to gate this feature on a block version change, so that clients have time to upgrade before it gets enabled and used. This is similar to what we have done in previous changes that affect data going into the ledger.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+We propose to add a third field to the struct, `multi_signers`, that allows including other `SignerSet`s. Following the change, the struct will look like this:
+```rust
+pub struct SignerSet<P: Default + PublicKey + Message> {
+    /// List of potential individual signers.
+    #[prost(message, repeated, tag = "1")]
+    #[digestible(name = "signers")]
+    individual_signers: Vec<P>,
+
+    /// List of potential sets of signers. This allows us to declare a nested
+    /// multi-signature scheme. Ideally the two arrays in this struct would be
+    /// combined into a single one that holds an enum of (IndividualSigner,
+    /// MultiSignerSet), but since this feature was added after we went live
+    /// and instances of SignerSet made their way into the ledger, it is not
+    /// possible to change the array type without breaking backwards
+    /// compatibility. This is also the reason the tag numbers in the struct
+    /// are not sequential.
+    #[prost(message, repeated, tag = "3")]
+    multi_signers: Vec<SignerSet<P>>,
+
+    /// Minimum number of signers required. The potential signers are the union
+    /// of `individual_signers` and `multi_signers`.
+    /// This implies that the upper limit (total number of possible signers) is
+    /// `individual_signers.len() + multi_signers.len()`.
+    #[prost(uint32, tag = "2")]
+    threshold: u32,
+}
+```
+
+The matching proto definitions inside `external.proto` will also need to be changed to include this new field.
+
+Since we are gating the support for the nested signer sets on a block version bump, there are places where we will need to ensure a `SignerSet` does not have any `multi_signers`, unless we are running a block version that supports it:
```

Comment:
:upside_down_face: 
```suggestion
Since we are gating the support for the nested multi-signatures on a block version bump, there are places where we will need to ensure a `SignerSet` does not have any `multi_signers`, unless we are running a block version that supports it:
```

Reply:
This is again the issue of signer-sets vs multi-signatures. The `MultiSig` struct didn't actually change, it's just the `SignerSet` that changed. So I am not sure whats better here.

Reply:
I agree, did a second review to try and clarify